java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestGenerationUtils.java

```diff
@@ -152,16 +163,280 @@ public static WorkspaceEdit generateTests(List<Object> arguments, IProgressMonit
+                .append(";")
+                .append(lineDelimiter)
+                .append(lineDelimiter);
+        }
+        buf.append(typeContent);
+        return buf.toString();
+    }
+
+    private static String constructTypeStub(ICompilationUnit testUnit, List<String> methods,
+            TestKind testKind, String lineDelimiter) throws CoreException {
+        final String typeName = testUnit.getElementName().replace(".java", "");
+        final StringBuilder buf = new StringBuilder();
+        buf.append("public class ").append(typeName).append(" {").append(lineDelimiter);
+        for (final String method : methods) {
+            buf.append(constructMethodStub(testUnit, testKind, method, lineDelimiter)).append(lineDelimiter);
+        }
+        buf.append("}").append(lineDelimiter);
+        return buf.toString();
+    }
+
+    private static String constructMethodStub(ICompilationUnit testUnit, TestKind testKind,
+            String method, String lineDelimiter) {
+        final StringBuilder buf = new StringBuilder();
+        buf.append("@Test").append(lineDelimiter);
+        if (testKind == TestKind.JUnit) {
+            buf.append("public ");
+        }
+        final String methodName = getTestMethodName(method);
+        buf.append("void ").append(methodName).append("() {").append(lineDelimiter).append(lineDelimiter).append("}");
+        final String methodContent = buf.toString();
+        // TODO: get test unit options directly
+        return CodeFormatterUtil.format(CodeFormatter.K_STATEMENTS, methodContent, 1,
+                lineDelimiter, testUnit.getJavaProject().getOptions(true));
+    }
+
+    private static IClasspathEntry getTestClasspathEntry(ICompilationUnit unit) throws JavaModelException {
+        final IJavaProject javaProject = unit.getJavaProject();
+        // In most cases, this is the classpath entry used for testing, we first find the target entry by hard-code
+        // to avoid go into the generated entries.
+        final IClasspathEntry testEntry = javaProject.getClasspathEntryFor(
+            javaProject.getPath().append("src/test/java"));
+        if (ProjectTestUtils.isTestEntry(testEntry)) {
+            return testEntry;
+        }
+
+        final IClasspathEntry[] entries = javaProject.readRawClasspath();
+        for (final IClasspathEntry entry : entries) {
+            if (ProjectTestUtils.isTestEntry(entry)) {
+                return entry;
+            }
+        }
+
+        return entries[0];
+    }
+
+    private static String getTestFullyQualifiedName(ITypeBinding typeBinding, IJavaProject project,
+            IClasspathEntry testEntry) throws JavaModelException {
+        String promptName = typeBinding.getBinaryName() + "Tests";
+        final ICompilationUnit testCompilationUnit = getTestCompilationUnit(project, testEntry, promptName);
+        if (!testCompilationUnit.exists()) {
+            promptName = typeBinding.getBinaryName() + "Test";
+        }
```

Comment:
I didn't get how it works.

Reply:
This is to get the default value of the fully-qualified name for the target test file.

It first check if `xxxTests.java` exists or not. If it does not exist, then will use `xxxTest` as the default name

Reply:
In addition to checking the corresponding test file with current source file, you should check if the package has any *Tests.java as well.

Reply:
ok